### ISSUE TITLE
[Question] How to use Attention to aggregate history?

### ISSUE THREAD

**Issue by hkristof03** [tensorflow/recommenders#641]
@OmarMAmin mentioned [Exploring Heterogeneous Metadata for Video Recommendation with
Two-tower Model](https://assets.amazon.science/1e/e6/4d7f8a2741a4a3b148e20a953946/exploring-heterogeneous-metadata-for-video-recommendation-with-two-tower-model.pdf) paper in [this discussion](https://github.com/tensorflow/recommenders/issues/633).

I read through the paper multiple times and a few resources about existing Attention mechanisms, but I haven't figured out the following questions:

1. Which Attention should be used from TF, [Luong-style attention](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Attention) or [Bahdanau-style attention.](https://www.tensorflow.org/api_docs/python/tf/keras/layers/AdditiveAttention)?
2. What should be exactly the query and value inputs? They train a User-Item recommender and it looks like that for the Attention layers in both the Query and Candidate Towers they use either User or Item features.

![image](https://user-images.githubusercontent.com/10969182/220878716-a4309183-644f-4b73-bc89-74a41b294dff.png)

From the paper:

![image](https://user-images.githubusercontent.com/10969182/220879828-471ef32c-0461-4a20-ae67-5cbbd5b1b126.png)

![image](https://user-images.githubusercontent.com/10969182/220879717-cbf6d65c-5a4f-4a9c-9eb8-dbc78ef8e8b3.png)

Maybe @patrickorlando do you have an idea related to this?

**Comment by patrickorlando** [comment]
Yeah I can't find any specifics in this paper @hkristof03, but I've seen the use of scaled dot-product attention (Luong Style) in other papers like [SASrec](https://arxiv.org/pdf/1808.09781.pdf). 

There are some useful implementations of the transformer encoder blocks in the keras-nlp package [here](https://github.com/NVIDIA-Merlin/models/blob/main/merlin/models/tf/core/encoder.py).

**Comment by caesarjuly** [comment]
I haven't looked at the paper yet. 
In my experience, if sequential history is essential, you can use self-attention or even Transformer encoder to learn the interactions and extract features. And then put an additive attention layer after the multiple output embeddings. It will aggregate them into one global embedding.
You can refer to the [implementation](https://github.com/microsoft/recommenders/blob/bbbed05593e81b53834b64a15549d4bec5e25875/recommenders/models/newsrec/models/layers.py#L10) of Microsoft teams.
In some places, it's also called attentive pooling.
![图片](https://user-images.githubusercontent.com/927521/225466128-34b2211b-ca82-40fa-ba84-0c18d6f75aae.png)
